upstairs/src/upstairs.rs

```diff
@@ -3504,4 +3508,641 @@ pub(crate) mod test {
+        up.apply(UpstairsAction::Downstairs(DownstairsAction::Client {
+            client_id: ClientId::new(0),
+            action: ClientAction::Response(Message::ReadResponse {
+                upstairs_id: up.cfg.upstairs_id,
+                session_id: up.cfg.session_id,
+                job_id: JobId(1000),
+                responses,
+            }),
+        }))
+        .await;
+        // no panic, great work everyone
+    }
+
+    #[tokio::test]
+    async fn bad_decryption_means_panic() {
+        let mut up = make_encrypted_upstairs();
+        up.force_active().unwrap();
+        set_all_active(&mut up.downstairs);
+
+        let data = Buffer::new(512);
+        let offset = Block::new_512(7);
+        let (tx, _rx) = oneshot::channel();
+        up.apply(UpstairsAction::Guest(BlockReq::new(
+            BlockOp::Read { offset, data },
+            tx,
+        )))
+        .await;
+
+        // fake read response from downstairs that will fail decryption
+        let mut data = Vec::from([1u8; 512]);
+
+        let (nonce, tag, _) = up
+            .cfg
+            .encryption_context
+            .as_ref()
+            .unwrap()
+            .encrypt_in_place(&mut data)
+            .unwrap();
+
+        let nonce: [u8; 12] = nonce.into();
+        let mut tag: [u8; 16] = tag.into();
+
+        // alter tag
+        if tag[3] == 0xFF {
+            tag[3] = 0x00;
+        } else {
+            tag[3] = 0xFF;
+        }
+
+        // compute integrity hash after alteration above! It should still
+        // validate
+        let hash = integrity_hash(&[&nonce, &tag, &data]);
+
+        let responses = Ok(vec![ReadResponse {
+            eid: 0,
+            offset,
+
+            data: BytesMut::from(&data[..]),
+            block_contexts: vec![BlockContext {
+                encryption_context: Some(
+                    crucible_protocol::EncryptionContext { nonce, tag },
+                ),
+                hash,
+            }],
+        }]);
+
+        // Prepare to receive the message with an invalid tag
+        let fut =
+            up.apply(UpstairsAction::Downstairs(DownstairsAction::Client {
+                client_id: ClientId::new(0),
+                action: ClientAction::Response(Message::ReadResponse {
+                    upstairs_id: up.cfg.upstairs_id,
+                    session_id: up.cfg.session_id,
+                    job_id: JobId(1000),
+                    responses,
+                }),
+            }));
+
+        let result = std::panic::AssertUnwindSafe(fut).catch_unwind().await;
+        assert!(result.is_err());
+        let r = result
+            .as_ref()
+            .unwrap_err()
+            .downcast_ref::<String>()
+            .unwrap();
+        assert!(
+            r.contains("DecryptionError"),
+            "panic for the wrong reason: {r}"
+        );
+    }
+
+    #[tokio::test]
+    async fn bad_hash_on_encrypted_read_panic() {
+        let mut up = make_encrypted_upstairs();
```

Comment:
Could we add a comment here on what makes this test different than the
`bad_read_hash_makes_panic` test that follows?  I think this test is altering
the tag?

Reply:
Good catch, this test was misleading (it altered the tag, but _also_ had a bad hash).  I removed the tag changes, since those are tested elsewhere, and added a comment.